elastic/kibana
# [Reporting] New UI for migrating reporting indices ILM policy

## Summary

This contribution adds new UI to the reporting plugin for guiding users to migrate their reporting indices to the reporting-provided ILM policy. Introduced in https://github.com/elastic/kibana/pull/103850.

## How to test

1. Start Kibana + ES on basic and activate your trial to access all features
2. In Kibana's start up logs you should see `Creating ILM policy for managing reporting indices...` if this is the first time starting with the new changes. Subsequent starts should contain a debug callout that this step is being skipped.
3. Navigate to the "Reporting" plugin, on a fresh start this page should be empty, no callout present.
4. Create a report by going to the "dashboard" app and using the "Share" menu to generate a PDF report
5. Navigate to the "Reporting" plugin, there should still be no callout present since all of your indices are up-to-date
6. Run the following request, unsetting the ILM policy managing reporting indices:

```bash
curl --request PUT \
  --url 'http://localhost:9200/.reporting-*/_settings' \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --data '{
	"settings": {
		"index.lifecycle.name": null
	}
}'
```

7. Go to the reporting plugin, the callout to migrate indices should be present (as well as the link to the `kibana-reporting` ILM policy in the bottom right), clicking it should migrate all reporting indices to use the reporting ILM policy.
8. Re-run the request to unset the ILM policies and delete the `kibana-reporting` ILM policy, can be done through the ILM policy UI
9. Navigate to the Reporting UI, this time the banner should be present _and_ the link to the ILM policy should be removed
10. Press the primary button in the call-to-action
11. The callout should be gone and the link to the ILM policy should be back (clicking this link takes you to the ILM UI)


### Notes

* There was jest snapshot that was captured against UI that was in "loading" state. The changes introduced here await the end of the loading state and so the snapshot is populated with a lot more UI than before. This is the primary source of new lines.


## Screenshots

### Call-to-action to migrate reporting indices

<img width="1649" alt="Screenshot 2021-06-29 at 15 55 30" src="https://user-images.githubusercontent.com/8155004/123938582-87689800-d997-11eb-86eb-1de4bc5e7e7f.png">

### Successful migration of reporting indices

* Note the success callout
* Note the link to "Edit ILM policy" in the bottom right

<img width="1677" alt="Screenshot 2021-06-29 at 15 55 56" src="https://user-images.githubusercontent.com/8155004/123938616-8df70f80-d997-11eb-979c-c6bb0ab5a9aa.png">

### ILM policy exists, but migration is still needed

<img width="843" alt="Screenshot 2021-06-29 at 15 56 27" src="https://user-images.githubusercontent.com/8155004/123938833-c696e900-d997-11eb-84f3-71a760013ae3.png">


### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| We could overwrite an existing ILM policy | Low | Med | This requires that a user has a policy called `kibana-reporting` and that there is a bug in our code that will re-create the policy. Further mitigations will be achieved by adding a functional API test |
